Userland/Libraries/LibDeviceTree/DeviceTree.cpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2024, Leon Albrecht <leon.a@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "DeviceTree.h"
+#include "FlattenedDeviceTree.h"
+#include <AK/NonnullOwnPtr.h>
+#include <AK/Types.h>
+
+namespace DeviceTree {
+
+ErrorOr<NonnullOwnPtr<DeviceTree>> DeviceTree::parse(ReadonlyBytes flattened_device_tree)
+{
+    // Device tree must be 4-byte aligned
+    // As the fields in it have to be 4-byte aligned
+    if ((bit_cast<FlatPtr>(flattened_device_tree.data()) & 3) != 0)
+        return Error::from_errno(EINVAL);
+
+    auto device_tree = TRY(adopt_nonnull_own_or_enomem(new (nothrow) DeviceTree { flattened_device_tree }));
+    DeviceTreeNodeView* current_node = device_tree.ptr();
+
+    // FIXME: Maybe dont assume the data is aligned
+    auto const& header = *reinterpret_cast<FlattenedDeviceTreeHeader const*>(flattened_device_tree.data());
+
+    TRY(walk_device_tree(header, flattened_device_tree,
+        {
+            .on_node_begin = [&current_node](StringView name) -> ErrorOr<IterationDecision> {
+                // Skip the root node, which has an empty name
+                if (name.is_empty())
+                    return IterationDecision::Continue;
+
+                // FIXME: Use something like children.emplace
+                TRY(current_node->children().try_set(name, DeviceTreeNodeView { current_node }));
+                auto& new_node = current_node->children().get(name).value();
+                current_node = &new_node;
+                return IterationDecision::Continue;
+            },
+            .on_node_end = [&current_node](StringView) -> ErrorOr<IterationDecision> {
+                current_node = current_node->parent();
```

Comment:
You should also maybe check if `name.is_empty()` here?

Reply:
We should, *should*, not exit another node after exiting the root node, so the `current_node` being NULL after this point should be fine